### ISSUE TITLE
Add rule for chunk-download-size-limits

### ISSUE THREAD

**Issue by icy0307** [statoscope/statoscope#203]
**Is your feature request related to a problem? Please describe.**
I try to limit the size of the critical path our application needs on ci.
However, our application is running in a host application(SPA), that serves several other  'apps' on the same page.
So `@statoscope/webpack/entry-download-size-limits` no longer suits us, cause we are no longer initial chunks.
Is there any other way to provide this kind of rule, that limits the size of the chunks that critical rendering path need?

**Describe the solution you'd like**
One way to achieve this is to specify the chunk that the rendering codes are in (like a designated webpack chunks name), then find all its parent chunks.

**Describe alternatives you've considered**
Another way of doing it is just to find all the chunks that the critical rendering path needs, simply just by its module name.
However, a module might be duplicated in different chunks, I am not very clear on how to implement this way(But isn't that rendering code should only appear once in the final builds?).

**Additional context**
I tried to implement a simple version of it by myself, following the [custom rule guide](https://github.com/statoscope/statoscope/tree/b02bbca5d4b4d3a262aafd7912ef27751dc92671/packages/stats-validator-plugin-webpack/src/rules).
But get stuck in compilations API in  ` WebpackRule<Params> = (ruleParams, data, api): void` of date object.
Although in my `stats.json`, the `parents` property exists for every chunk, I can't get it from 'data.compilations[0].chunks[someIndex].parents'.
Looking into the source code, 
I found in the [source code](https://github.dev/statoscope/statoscope/blob/master/packages/webpack-model/src/handleCompilation.ts#L531), the chunk's parent has been set to empty and relink by chunks' origin, which is not in ours `stats.json`
1. What is your idea on implementing this properly?
2. Is it ok we use just js to query the original compilation object instead of jora? I and many of my teams find it quite hard to learn `jora` by simply looking into the demo. (Is there any tutorial on this subject?) .Would you accept the rule's pr without `jora`?
3. why is chunks origin is needed for this situation?

**Comment by smelukov** [comment]
Thanks for the issue!

> What is your idea on implementing this properly?

Let's start from rule parameters:
```ts
export type Params = {
  byName: Array<{name: string | RegExp; limit: number; useCompressed: boolean}>
};
```

So, here is a pseudo-implementation of the rule:

```ts
const chunkDownloadSizeLimits: WebpackRule<Params> = (params, data, api): void => {
  api.setRuleDescriptor(...);
  
  for (const item of params.byName) {
    const targetChunks = getTargetChunksFromDataWithParams(data, params); // collect chunks by name or regexp (you could use jora or pure js here)
    for (const chunk of targetChunks) {
      const diff = isChunkSizeOk(chunk, item); // check chunk size by the item props
      if (diff) {
        api.message('Oh, no!', {
          filename: data.files[0].name,
          compilation: chunk.compilation.hash,
          related: [{ type: 'chunk', id: chunk.id }],
        });
      }
    }
  }
};

export default chunkDownloadSizeLimits;
```

> Is it ok we use just js to query the original compilation object instead of jora?

Yep

> why is chunks origin is needed for this situation?

Chunks origins is more accuracy to get chunk relations from webpack stats

**Comment by icy0307** [comment]
Thanks for your detailed answer.
1. Can I find all the files by module name? But isn't that one module could belong to several chunks?
2.  Could you elaborate on why `Chunks origins is more accuracy to get chunk relations from webpack stats`?

One more thing, the `stats.json` our project generated is quite large (around 1.5GB before zipped), is there anything I could to about it?

**Comment by smelukov** [comment]
> Can I find all the files by module name?

Yep. Every module includes chunks-property. You could find all the modules you need (e.g. by module name) and then collect chunks from the chunks-property.

> But isn't that one module could belong to several chunks?

Yes, one module could belong to several chunks. That's why chunks-property is an array

> Could you elaborate on why Chunks origins is more accuracy to get chunk relations from webpack stats?

Seems like there is a bug in webpack stats (see #168 for more info). Do you have concerns with using origins to build chunk graph?
